Correctly handle deeply nested scopes

trescenzi opened the pull request:
When large batch DOM updates are made the document mutation observer that handles rescoping things will fire with only the root most node of the newly attached tree. This is expected behavior of mutation observer. As a result it's possible that multiple shadow roots will be appended in a single tick of the observer with unscoped children. Previously every unscoped child would be scoped to the scope of the root most scope. However this can be incorrect when multiple shadow boundaries have been appended in a tree. In order to prevent incorrectly scoping children nodes before scoping them their root and host need to be identified. The correct scope will then be applied instead of assuming that it's the scope of the newly added node.

This fixes: Polymer/polymer#4910 in addition to the intent behind #162. 

[This reproducer](https://codepen.io/trescenzi/pen/bvXxMG?editors=1010) shows, as simply as I could, the root problem. It also shows the necessity of checking for unscoped children whenever the mutation observer is called, not just when the scope being updated is the same as before(`newScope === currentScope`). However I'm concerned that this will be a performance hit so I'm not sure that this is the best solution.

If I'm missing a way to correctly use ShadyCSS please let me know. I've tried using `styleElement` and `styleSubtree` but it didn't seem to solve the problem. It looks like the intent of the MutationObserver is to handle situations like these where unscoped, and incorrectly scoped, elements are added dynamically into the DOM so I figured addressing the problem here was best.

Thank you!

trescenzi commented:
Working on a test and I'm pretty consistently getting fails locally that go away when I remove my change. Most likely it's related to the removal of the `continue`. I'll fix those but I wanted to mention that I was having failures while the build passed in CI. It looks like no build runs in CI. Am I expected to check in the built files with this commit?

TimvdLippe commented:
Yes you need to run `npm run build` and test with that bundle. It is required for CI to make sure we can run the code properly on very old browsers.

trescenzi commented:
Updated to fix the failing tests and added a test to the dynamic scope tests. It's a little more complex since I couldn't use `makeElement` as it waits for `connectedCallback` to attach a shadow root and children which then paves over the problem.

trescenzi commented on src/document-watcher.js:
```diff
@@ -71,19 +71,32 @@ function handler(mxns) {
-            StyleTransformer.element(unscoped[j], currentScope);
+        // make sure all the subtree elements are scoped correctly
+        let unscoped = window['ShadyDOM']['nativeMethods']['querySelectorAll'].call(
+          n, `:not(.${StyleTransformer.SCOPE_NAME})`);
```
As far as I can tell that's the core issue that I was running into. Because this is run in a mutationObserver when a large tree is inserted only the root of that tree will be passed into this function as an `addedNode` in the mutation list. Previously this was selecting everything and applying the root's scope, which is definitely incorrect since this goes past scope boundaries. So instead of updating all unscoped nodes to the root's scope this grabs all unscoped nodes, finds their correct scope, and scopes them.

I am concerned though about the fact that we're digging so deeply into the tree as you're saying. It seems dangerous to grab all of the nodes in the tree like this and run over them. I *think* the safety comes in that we're only grabbing nodes that don't have `.style-scope` on them so we won't touch anything that's already been scoped. And those that haven't been scoped should be scoped.

TimvdLippe replied:
Oh I see now. Yes `getIsExtends(unscopedHost)` makes more sense now. Maybe we can improve the naming a little bit to clarify?

What about:
`s/unscoped/usncopedNodes/`
`s/unscopedRoot/rootForUnscopedNode/`
`s/unscopedHost/hostForUnscopedNode/`
`s/newScope/scopeForPreviouslyUnscopedNode/`

And then maybe add another comment saying:
```
To make sure that any unscoped nodes that were inserted in the current batch are correctly styled,
query all unscoped nodes and force their style-scope to be applied.
This could happen if a sub-element appended an unscoped node in its shadowroot and this function
runs on a parent element of the host of that unscoped node:
parent-element -> element -> unscoped node
Here unscoped node should have the style-scope element, not parent-element.
```

trescenzi commented:
Hi @TimvdLippe @azakus any updates on this?

TimvdLippe commented:
@azakus Is rather busy at the moment, so reviewing is a bit delayed. Hopefully he can get to this pr soonTM.

trescenzi commented:
Any updates on this? I can fix the conflicts whenever just let me know and I'll get it back up to date.

azakus commented:
Sorry for the delay in reviewing this. The solution looks good to me, I'll handle the merge conflicts and get this in.

azakus commented:
@trescenzi Actually, could you sign the Google CLA really quick: https://cla.developers.google.com/ ?

andreasanheuser commented:
Ping @azakus, @trescenzi, can we get this in?

trescenzi commented:
Just signed the CLA. Let me know if I need to do anything else @azakus. Thanks again!

TimvdLippe commented:
@trescenzi could you resolve the merge conflicts?

trescenzi commented:
I've resolved the conflicts.

TimvdLippe commented:
Thanks!
